Python 3 incompatibilities

[flake8](http://flake8.pycqa.org) testing of https://github.com/lihanghang/Knowledge-Graph on Python 3.7.1

$ __flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics__
```
./Algorithm-code/RelationExtraction_CNN+ATT/main_att.py:24:35: E999 TabError: inconsistent use of tabs and spaces in indentation
		torch.cuda.set_device(opt.gpu_id)
                                  ^
./Algorithm-code/RelationExtraction_CNN+ATT/utils.py:19:22: E999 TabError: inconsistent use of tabs and spaces in indentation
        fp_out.close()
                     ^
./Algorithm-code/RelationExtraction_CNN+ATT/datasets/filternyt.py:87:55: F821 undefined name 'file'
        wordlist.extend([word.strip('\n') for word in file(self.word_path)])
                                                      ^
./Algorithm-code/RelationExtraction_CNN+ATT/datasets/filternyt.py:89:21: F821 undefined name 'file'
        for line in file(self.w2v_path):
                    ^
./Algorithm-code/RelationExtraction_CNN+ATT/datasets/filternyt.py:112:13: F821 undefined name 'file'
        f = file(path)
            ^
./Algorithm-code/RelationExtraction_CNN+ATT/models/PCNN_ATT.py:96:34: E999 TabError: inconsistent use of tabs and spaces in indentation
	def init_int_constant(self, num):
                                 ^
./Algorithm-code/Text-Similarity/train.py:54:34: E999 SyntaxError: invalid syntax
print '训练开始......................'
                                 ^
4     E999 TabError: inconsistent use of tabs and spaces in indentation
3     F821 undefined name 'file'
7
```
__E901,E999,F821,F822,F823__ are the "_showstopper_" [flake8](http://flake8.pycqa.org) issues that can halt the runtime with a SyntaxError, NameError, etc. These 5 are different from most other flake8 issues which are merely "style violations" -- useful for readability but they do not effect runtime safety.
* F821: undefined name `name`
* F822: undefined name `name` in `__all__`
* F823: local variable name referenced before assignment
* E901: SyntaxError or IndentationError
* E999: SyntaxError -- failed to compile a file into an Abstract Syntax Tree

--- a/Algorithm-code/RelationExtraction_CNN+ATT/main_att.py
+++ b/Algorithm-code/RelationExtraction_CNN+ATT/main_att.py
@@ -10,7 +10,6 @@ from utils import save_pr, now, eval_metric
 
 
 def collate_fn(batch):
-
 	data, label = zip(*batch)
 	return data, label
 
@@ -18,44 +17,44 @@ def test(**kwargs):
 	pass
 
 def train(**kwargs):
-    kwargs.update({'model': 'PCNN_ATT'})
-    opt.parse(kwargs)
-    if opt.use_gpu:
+	kwargs.update({'model': 'PCNN_ATT'})
+	opt.parse(kwargs)
+	if opt.use_gpu:
 		torch.cuda.set_device(opt.gpu_id)
 
-    model = getattr(models, 'PCNN_ATT')(opt)
-    if opt.use_gpu:
+	model = getattr(models, 'PCNN_ATT')(opt)
+	if opt.use_gpu:
 		model.cuda()
 
 	# loading data
-    DataModel = getattr(datasets, opt.data + 'Data')
-    train_data = DataModel(opt.data_root, train=True)
-    train_data_loader = DataLoader(train_data, opt.batch_size, shuffle=True, num_workers=opt.num_workers, collate_fn = collate_fn)
-    
-    test_data = DataModel(opt.data_root, train=False)
-    test_data_loader = DataLoader(test_data, opt.batch_size, shuffle=False, num_workers=opt.num_workers, collate_fn = collate_fn)
-    print('{} train data: {}; test data: {}'.format(now(), len(train_data), len(test_data)))
+	DataModel = getattr(datasets, opt.data + 'Data')
+	train_data = DataModel(opt.data_root, train=True)
+	train_data_loader = DataLoader(train_data, opt.batch_size, shuffle=True, num_workers=opt.num_workers, collate_fn = collate_fn)
 
-    optimizer = optim.Adadelta(model.parameters(), rho=0.95, eps=1e-6)
+	test_data = DataModel(opt.data_root, train=False)
+	test_data_loader = DataLoader(test_data, opt.batch_size, shuffle=False, num_workers=opt.num_workers, collate_fn = collate_fn)
+	print('{} train data: {}; test data: {}'.format(now(), len(train_data), len(test_data)))
+
+	optimizer = optim.Adadelta(model.parameters(), rho=0.95, eps=1e-6)
 
 
-    # train
-    for epoch in range(opt.num_epochs):
+	# train
+	for epoch in range(opt.num_epochs):
 		total_loss = 0
 		for idx, (data, label_set) in enumerate(train_data_loader):
 
-		    label = [l[0] for l in label_set]
+			label = [l[0] for l in label_set]
 
-		    optimizer.zero_grad()
-		    model.batch_size = opt.batch_size
-		    loss = model(data, label)
-		    if opt.use_gpu:
-		        label = torch.LongTensor(label).cuda()
-		    else:
-		        label = torch.LongTensor(label)
-		    loss.backward()
-		    optimizer.step()
-		    total_loss += loss.item()
+			optimizer.zero_grad()
+			model.batch_size = opt.batch_size
+			loss = model(data, label)
+			if opt.use_gpu:
+				label = torch.LongTensor(label).cuda()
+			else:
+				label = torch.LongTensor(label)
+			loss.backward()
+			optimizer.step()
+			total_loss += loss.item()
 
 		if epoch > 2:
 			pred_res, p_num = predict_var(model, test_data_loader)
@@ -145,6 +144,3 @@ def predict(model, test_data_loader):
 if __name__ == '__main__':
 	import fire
 	fire.Fire()
-		
-
-

--- a/Algorithm-code/RelationExtraction_CNN+ATT/models/PCNN_ATT.py
+++ b/Algorithm-code/RelationExtraction_CNN+ATT/models/PCNN_ATT.py
@@ -8,15 +8,12 @@ import torch.nn.functional as F
 from torch.autograd import Variable
 
 
-
-
 '''
 References:
 《Neural Relation Extraction with Selective Attention over Instances》--PCNN+ATT(2016)
 《Distant Supervision for Relation Extraction via Piecewise　Convolutional Neural Networks》--PCNN(2015)
 '''
 class PCNN_ATT(BasicModule):
-	
 
     def __init__(self, opt):
         super(PCNN_ATT, self).__init__()
@@ -62,21 +59,19 @@ class PCNN_ATT(BasicModule):
     def init_model_weight(self):
         """
          here use xavier approach
-        """ 
+        """
         nn.init.xavier_uniform(self.rel_embs)
         nn.init.uniform(self.rel_bias)
         for conv in self.convs:
-        	nn.init.xavier_uniform(conv.weight)
-        	nn.init.uniform(conv.bias)
-
-
+            nn.init.xavier_uniform(conv.weight)
+            nn.init.uniform(conv.bias)
 
     def init_word_emb(self):
         def p_2norm(path):
             v = torch.from_numpy(np.load(path))
             if self.opt.norm_emb:
-            	v = torch.div(v, v.norm(2 ,1).unsqueeze(1))
-            	v[v != v] == 0.0
+                v = torch.div(v, v.norm(2 ,1).unsqueeze(1))
+                v[v != v] == 0.0
             return v
 
         w2v = p_2norm(self.opt.w2v_path)
@@ -84,23 +79,23 @@ class PCNN_ATT(BasicModule):
         p2_2v = p_2norm(self.opt.p2_2v_path)
 
         if self.opt.use_gpu:
-        	self.word_embs.weight.data.copy_(w2v.cuda())
-        	self.pos1_embs.weight.data.copy_(p1_2v.cuda())
-        	self.pos2_embs.weight.data.copy_(p2_2v.cuda())
+            self.word_embs.weight.data.copy_(w2v.cuda())
+            self.pos1_embs.weight.data.copy_(p1_2v.cuda())
+            self.pos2_embs.weight.data.copy_(p2_2v.cuda())
         else:
-        	self.pos1_embs.weight.data.copy_(p1_2v)
-        	self.pos2_embs.weight.data.copy_(p2_2v)
-        	self.word_embs.weight.data.copy_(w2v)
+            self.pos1_embs.weight.data.copy_(p1_2v)
+            self.pos2_embs.weight.data.copy_(p2_2v)
+            self.word_embs.weight.data.copy_(w2v)
 
 
-	def init_int_constant(self, num):
-		'''
-        a util function for generating a LongTensor Variable
-		'''
-		if self.opt.use_gpu:
-			return Variable(torch.LongTensor([num]).cuda())
-		else:
-			return Variable(torch.LongTensor([num]))
+        def init_int_constant(self, num):
+            '''
+    a util function for generating a LongTensor Variable
+            '''
+            if self.opt.use_gpu:
+                return Variable(torch.LongTensor([num]).cuda())
+            else:
+                return Variable(torch.LongTensor([num]))
 
 
     def mask_piece_pooling(self, x, mask):
@@ -125,87 +120,85 @@ class PCNN_ATT(BasicModule):
         batch_res = []
 
         for i in range(split_pool):
-        	ins = split_batch_x[i].squeeze()
-        	pool = split_pool[i].squeeze().data
-        	seg_1 = ins[:, :pool[0]].max(1)[0].unsqueeze(1)
-        	seg_2 = ins[:, pool[0]: pool[1]][0].unsqueeze(1)
-        	seg_3 = ins[:, pool[1]:].max(1)[0].unsqueeze(1)
-        	piece_max_pool = torch.cat([seg_1, seg_2, seg_3], 1).view(1, -1)
-        	batch_res.append(piece_max_pool)
+            ins = split_batch_x[i].squeeze()
+            pool = split_pool[i].squeeze().data
+            seg_1 = ins[:, :pool[0]].max(1)[0].unsqueeze(1)
+            seg_2 = ins[:, pool[0]: pool[1]][0].unsqueeze(1)
+            seg_3 = ins[:, pool[1]:].max(1)[0].unsqueeze(1)
+            piece_max_pool = torch.cat([seg_1, seg_2, seg_3], 1).view(1, -1)
+            batch_res.append(piece_max_pool)
 
         out = torch.cat(batch_res, 0)
         assert out.size(1) == 3 * self.opt.filters_num
         return out
 
-
-
     def forward(self, x, label=None):
-    	# get all sentences embedding in all bags of one batch
-    	self.bags_feature = self.get_bags_features(x)
+        # get all sentences embedding in all bags of one batch
+        self.bags_feature = self.get_bags_features(x)
 
-    	if label is None:
-    		# for test
-    		assert self.training is False
-    		return self.test(x)
-    	else:
-    		# for train
-    		assert self.training is True
-    		return self.fit(x, label)
+        if label is None:
+            # for test
+            assert self.training is False
+            return self.test(x)
+        else:
+            # for train
+            assert self.training is True
+            return self.fit(x, label)
 
 
     def fit(self, x, label):
-    	'''
-    	train process
-    	'''
+        '''
+        train process
+        '''
         x = self.get_batch_feature(label)
         x = self.dropout(x)
         out = x.mm(self.rel_embs.t()) + self.rel_bias
         if self.opt.use_gpu:
-        	v_label = torch.LongTensor(label).cuda()
+            v_label = torch.LongTensor(label).cuda()
         else:
-        	v_label = torch.LongTensor(label)
+            v_label = torch.LongTensor(label)
         ce_loss = F.cross_entropy(out, Variable(v_label))
         return ce_loss
 
 
     def test(self, x):
-    	'''
+        '''
         test process
-    	'''
-    	pre_y = []
-    	for label in range(0, self.opt.rel_num):
-    		lables = [lables for _ in range(len(x))]
-    		bags_feature = self.get_bathch_feature(lables)
-    		out = self.test_scale_p * bags_feature.mm(self.rel_embs.t()) + self.rel_bias
-    		pre_y.append(out.unsqueeze(1))
+        '''
+        pre_y = []
+        for label in range(0, self.opt.rel_num):
+            lables = [label for _ in range(len(x))]
+            bags_feature = self.get_bathch_feature(lables)
+            out = self.test_scale_p * bags_feature.mm(self.rel_embs.t()) + self.rel_bias
+            pre_y.append(out.unsqueeze(1))
         res = torch.cat(pre_y, 1).max(1)[0]
         return F.softmax(res, 1).t()
 
 
     def get_batch_feature(self, lables):
-    	'''
+        '''
          Using Attention to get all bags embedding in a batch
-    	'''
-    	batch_feature = []
-    	for bag_embs, label in zip(self.bags_feature, lables):
-    		alpha = bag_embs.mm(self.rel_embs[label].view(-1, 1))
-    		bag_embs = bag_embs * F.softmax(alpha, 0)
-    		bag_vec = torch.sum(bag_embs, 0)
-    		batch_feature.append(bag_vec.unsqueeze(0))
+        '''
+        batch_feature = []
+        for bag_embs, label in zip(self.bags_feature, lables):
+            alpha = bag_embs.mm(self.rel_embs[label].view(-1, 1))
+            bag_embs = bag_embs * F.softmax(alpha, 0)
+            bag_vec = torch.sum(bag_embs, 0)
+            batch_feature.append(bag_vec.unsqueeze(0))
 
-    	return torch.cat(batch_feature, 0)
+        return torch.cat(batch_feature, 0)
 
 
     def get_bags_features(self, bags):
         '''
-        get all bags embedding in one batch before Attention 
-        '''   
+        get all bags embedding in one batch before Attention
+        '''
         bags_feature = []
         for bag in bags:
             if self.opt.use_gpu:
                 data = map(lambda x: Variable(torch.LongTensor(x).cuda()), bag)
             else:
-            	data = map(lambda x: Variable(torch.LongTensor(x)), bag)
+                data = map(lambda x: Variable(torch.LongTensor(x)), bag)
             bag_embs = self.get_ins_emb(data)
             bags_feature.append(bag_embs)
 
@@ -213,22 +206,19 @@ class PCNN_ATT(BasicModule):
 
 
     def get_ins_emb(self, x):
-    	'''
+        '''
         x: all instance in a Bag
-    	'''
-    	insEnt, _, insX, insPFs, insPool, mask = x
-    	insPF1, insPF2 = [i.squeeze(1) for i in torch.split(insPFs, 1, 1)]
+        '''
+        insEnt, _, insX, insPFs, insPool, mask = x
+        insPF1, insPF2 = [i.squeeze(1) for i in torch.split(insPFs, 1, 1)]
 
-    	word_emb = self.word_embs(insX)
-    	pf1_emb = self.pos1_embs(insPF1)
-    	pf2_emb = self.pos2_embs(insPF2)
+        word_emb = self.word_embs(insX)
+        pf1_emb = self.pos1_embs(insPF1)
+        pf2_emb = self.pos2_embs(insPF2)
 
-    	x = torch.cat([word_emb, pf1_emb, pf2_emb], 2)
-    	x = x.unsqueeze(1)
-    	x = [conv(x).squeeze(3) for conv in  self.convs]
-    	x = [self.mask_piece_pooling(i, mask) for i in x]
-    	x = torch.cat(x, 1).tanh()
-    	return x
-
-
-        
+        x = torch.cat([word_emb, pf1_emb, pf2_emb], 2)
+        x = x.unsqueeze(1)
+        x = [conv(x).squeeze(3) for conv in  self.convs]
+        x = [self.mask_piece_pooling(i, mask) for i in x]
+        x = torch.cat(x, 1).tanh()
+        return x

--- a/Algorithm-code/RelationExtraction_CNN+ATT/utils.py
+++ b/Algorithm-code/RelationExtraction_CNN+ATT/utils.py
@@ -16,7 +16,7 @@ def save_pr(out_dir, name, epoch, pre, rec, fp_res=None, opt=None):
 		fp_out = open('{}/{}_{}_FP.txt'.format(out_dir, name, epoch + 1), 'w')
 		for idx, r, p in fp_res:
 			fp_out.write('{}{}{}\n'.format(idx, r, p))
-        fp_out.close()
+	fp_out.close()
 
 	for p, r in zip(pre, rec):
 		out.write('{} {}\n'.format(p, r))

--- a/Algorithm-code/RelationExtraction_CNN+ATT/utils.py
+++ b/Algorithm-code/RelationExtraction_CNN+ATT/utils.py
@@ -61,10 +61,10 @@ def eval_metric(true_y, pred_y, pred_p):
 			precision = 1.0
 		else:
 			precision = tp * 1.0 / (fp + tp)
-        recall = tp * 1.0 / positive_num
-        if precision != all_pre[-1] or recall != all_rec[-1]:
-        	all_pre.append(precision)
-        	all_rec.append(recall)
+	recall = tp * 1.0 / positive_num
+	if precision != all_pre[-1] or recall != all_rec[-1]:
+		all_pre.append(precision)
+		all_rec.append(recall)
 
 
 	print("tp={}; fp={}; fn={}; positive_num={}".format(tp, fp, fn, positive_num))

